src/python/pants/backend/docker/target_types.py

```diff
@@ -28,9 +28,13 @@ class DockerBuildArgsField(StringSequenceField):
 
 
 class DockerImageSourceField(SingleSourceField):
-    default = "Dockerfile"
+    expected_num_files = range(0, 2)
```

Comment:
I don't think either of these changes are actually necessary? The validation only happens when the `sources` were explicitly declared, not when code generation happens to generate into this sources type. It continues to be true that we expect the user to declare exactly one file with a default of `Dockerfile`.

It happens to be the case that there's another way to get something that looks like `DockerImageSourceField`, that you can use a `dockerfile` target. Note that its codegen isn't actually creating an instance of `DockerImageSourceField` though, it only is returning back a `Digest` of something that represents the intent of `DockerImageSourceField`. Does that make sense?

An example: `relocated_files` is using codegen to do codegen against `FileSourceField`, even though it itself has 0 source files.

Reply:
I've observed this to be true.
When hydrating sources, the validation takes place for the `source` field with the default value if not explicitly declared in the BUILD file. So, if `default` is not empty, I'll get a glob mismatch against the workspace about not finding any matching sources, and the `expected_num_files` will err out if there was no source hydrated from the workspace.

I would prefer if there was a way to have the codegen sources be considered the same as workspace sources here.. but the codegen is invoked after the validation: 
https://github.com/pantsbuild/pants/blob/a103139bce1178e85e764f935270d155f7fa8c3c/src/python/pants/engine/internals/graph.py#L792-L805

Reply:
That is, if the field has a default value, I can't tell the difference between if it was declared explicitly, or if it is the default value.

Reply:
> if the field has a default value, I can't tell the difference between if it was declared explicitly, or if it is the default value.

I'm not following this. Why do you need to do that? Your rules to build and publish docker images should be agnostic to how the source file came about, whether it was generated vs. read from disk (via default or hardcoded value), right?

> So, if default is not empty, I'll get a glob mismatch against the workspace about not finding any matching sources, and the expected_num_files will err out if there was no source hydrated from the workspace.

Only for the `docker_image` target though, right? That target reads its Dockerfile from disk and its `source` field must be for an actual file, right? Regardless of if it's the default or explicitly declared.

Reply:
> I'm not following this. Why do you need to do that?

True. I don't want to.. the reasoning behind that train of thought is just me trying to figure out how to get around the next question.

> That target reads its Dockerfile from disk and its source field must be for an actual file, right?

Exactly the point. So if the `source` field for `docker_image` defaults to `Dockerfile`, we either have to provide that file, or override the field value. When using `dockerfile` target to generate the Dockerfile, I'd prefer not to have to also override the `source` field, in order for it to not err out on a missing `Dockefile` in the workspace. Makes sense?

Reply:
Hm, are you confusing target generation vs codegen perhaps? 

If you were doing target generation, where `dockerfile` generates a `docker_image`, then you _would_ need to have a Dockerfile written to disk so that the `docker_image`'s `source` field worked. Or, you could remove its validation.

Instead, you are using codegen. A `dockerfile` target is distinct, and all it does is create a synthetic `Dockerfile` that "looks like" `DockerImageSource` but isn't actually that. `expected_num_files` validation never runs on the generated files etc.

You should be able to land this change without making any changes at all to `docker_image` fwict.

Reply:
I'll take this in DM...

Reply:
FTR,  Found out that the input to `parse_dockerfile` should not be a source field, but rather an address instead. This makes the rule more generic in terms of what it can work with. Update coming...

Reply:
This change is now reverted.

Reply:
And back again. In order to facilitate using a dependency to generate the Dockerfile, the `source` field is now optional, so no default value.